### PR TITLE
flow: consul and apache integrations

### PR DESCRIPTION
**Pull request by captncraig**
Also moved a bunch of the code from node_exporter into a common package, much like discovery. Now all a new integration has to do is create a config type with a convert function, and wire it into the shared component.

### CODE REVIEW

File: component/prometheus/integration/integrationComponent.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	integration, err := c.creator(c.opts, args)
+	if err != nil {
+		return err
+	}
+	c.mut.Lock()
+	c.integration = integration
+	c.mut.Unlock()
+	c.reload <- struct{}{}
```
Comment (rfratto) [inline component/prometheus/integration/integrationComponent.go:98]: Looks good, only have one drive-by comment:

```suggestion
	select {
	case c.reload <- struct{}{}:
	default:
	}
```

File: component/prometheus/integration/integrationComponent.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		// Call to Update() to set the output once at the start.
+		if err := c.Update(args); err != nil {
+			return nil, err
+		}
+		targets := []discovery.Target{{
+			model.AddressLabel:     opts.HTTPListenAddr,
+			model.SchemeLabel:      "http",
+			model.MetricsPathLabel: path.Join(opts.HTTPPath, "metrics"),
+			"name":                 name,
+		}}
```
Comment (rfratto) [inline component/prometheus/integration/integrationComponent.go:53]: Reminder about #2424, it's not a blocker for this PR but it's still an open question for these integrations.

Comment (captncraig) [reply component/prometheus/integration/integrationComponent.go:53]: Thanks for reminder. I will think it over and make sure we finalize before merging.

File: component/all/all.go [hunk 3]
```diff
@@ -40,6 +40,7 @@ import (
 	_ "github.com/grafana/agent/component/otelcol/receiver/zipkin"              // Import otelcol.receiver.zipkin
 	_ "github.com/grafana/agent/component/phlare/scrape"                        // Import phlare.scrape
 	_ "github.com/grafana/agent/component/phlare/write"                         // Import phlare.write
+	_ "github.com/grafana/agent/component/prometheus/integration/consul"        // Import prometheus.integration.consul
 	_ "github.com/grafana/agent/component/prometheus/integration/node_exporter" // Import prometheus.integration.node_exporter
```
Comment (marctc) [inline component/all/all.go:44]: apache needs to be included too, right?

File: component/prometheus/integration/integrationComponent.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
+package integration
```
Comment (marctc) [inline component/prometheus/integration/integrationComponent.go:1]: Could this code also be used for non prometheus integrations (eventhandler)? also nit: use `_` or `-` for a filename instead of camel case?

**Review by marctc — COMMENTED** [review]
Looking good. We probably should add tests. For integrations should be enough to test conversion is properly done I guess

**Review by jcreixell — APPROVED** [review]
nice, 2 in 1, LGTM. The only question I have is if we want to have a consistent way of testing the configuration mapping (Convert function, in this case it is simple enough, in other cases it may be more complex).

If we don't want to test it and it isn't part of an interface, we may as well make it private

File: component/prometheus/integration/integrationComponent.go (hunk 4 above)
Comment (captncraig) [reply component/prometheus/integration/integrationComponent.go]: Yes, it possibly could. We may need to add additional extension points for that, but it could be done.

File: docs/sources/flow/reference/components/prometheus.integration.apache.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+```
+
+## Arguments
+The following arguments can be used to configure the exporter's behavior.
+All arguments are optional. Omitted fields take their default values.
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`scrape_uri`    | `string` | URI to apache stub status page. | `http://localhost/server-status?auto` | no
```
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.apache.md:29]:
```suggestion
`scrape_uri`    | `string` | URI to Apache stub status page. | `http://localhost/server-status?auto` | no
```

File: docs/sources/flow/reference/components/prometheus.integration.apache.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+## Exported fields
+The following fields are exported and can be referenced by other components.
+
+Name      | Type                | Description
+--------- | ------------------- | -----------
+`targets` | `list(map(string))` | The targets that can be used to collect `consul` metrics.
+
+For example, the `targets` could either be passed to a `prometheus.relabel`
+component to rewrite the metrics' label set, or to a `prometheus.scrape`
+component that collects the exposed metrics.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/prometheus.integration.apache.md:42]:
```suggestion
For example, the `targets` can either be passed to a `prometheus.relabel`
component to rewrite the metric's label set, or to a `prometheus.scrape`
component that collects the exposed metrics.
```

Comment (karengermond) [reply docs/sources/flow/reference/components/prometheus.integration.apache.md:42]: I'm unsure of this one.  It's the label set of a particular metric, right? So it would be "metric's"...

**Review by karengermond — APPROVED** [review]
LGTM!
